src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,363 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "uiScreens.h"
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+#define TRACE_WITH_CTX(message, ...) TRACE(message "level = %u, remaining scripts = %u", ##__VA_ARGS__, ctx->level, ctx->complexScripts[ctx->level].remainingScripts);
+#define SAFE_SPRINTF(str, format, ...) {\
+		explicit_bzero(str, SIZEOF(str));\
+		snprintf(str, SIZEOF(str), format, __VA_ARGS__);\
+		ASSERT(strlen(str) + 1 < SIZEOF(str));\
+	}
+
+static void expectsMoreScripts()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	VALIDATE(ctx->complexScripts[ctx->level].remainingScripts > 0, ERR_INVALID_STATE);
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+}
+
+static inline void complexScriptFinished()
+{
+	ASSERT(ctx->level > 0);
+	ctx->level--;
+
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("complex script finished, ");
+}
+
+// UI
+
+static void deriveScriptHash_display_ui_callback() {
+	io_send_buf(SUCCESS, NULL, 0);
```

Comment:
Take a look at `respondSuccessEmptyMsg()`. For some reasons that go fairly deep into very ledger-specific code, and I don't understand well, after io_send_buf, you want to also tell the UI something, e.g ui_displayBusy().